### PR TITLE
Numaflow Controller Rollout Auto Healing

### LINKED ISSUE #19: auto heal for NumaflowControllerRollout
Watch() NumaflowController Deployment and related resources to trigger Reconcile() like we do with PipelineRollout and ISBServiceRollout. https://github.com/numaproj/numaplane/issues/6 is an improvements we can consider later.

### PR DESCRIPTION
**Pull request by afugazzotto**
Fixes #19 

### Modifications

Enabled watching of Deployment resource in Numaflow Controller Rollout controller and updated unit test to test reconciliation logic.

### Verification

Tested manually and via unit tests that making changes to the Deployment resource owned by the NumaflowControllerRollout will auto heal by undoing the changes.

### CODE REVIEW

**Comment by afugazzotto** [comment]
Need to fix merge conflict.

File: internal/controller/numaflowcontrollerrollout_controller.go [hunk 1]
```diff
@@ -352,38 +368,27 @@ func (r *NumaflowControllerRolloutReconciler) getResourceOperations() (kubeUtil.
...
+	// Watch NumaflowControllerRollouts
+	if err := controller.Watch(source.Kind(mgr.GetCache(), &apiv1.NumaflowControllerRollout{}), &handler.EnqueueRequestForObject{}, predicate.GenerationChangedPredicate{}); err != nil {
+		return err
+	}
 
-	   		handler.EnqueueRequestForOwner(mgr.GetScheme(), mgr.GetRESTMapper(), &apiv1.NumaflowControllerRollout{}, handler.OnlyControllerOwner()),
-	   		predicate.GenerationChangedPredicate{}); err != nil {
-	   		return err
-	   	}
+	// Watch Deployments of numaflow-controller (NOTE: could watch other resources as well)
```
Comment (xdevxy) [inline internal/controller/numaflowcontrollerrollout_controller.go:381]: nit: this comment is not quite right? as we are watching `numaflow-controller `deployment only?

**Review by juliev0 — APPROVED** [review]

File: internal/controller/suite_test.go [hunk 2]
```diff
@@ -164,3 +198,23 @@ func downloadCRD(url string, downloadDir string) {
 	_, err = io.Copy(out, resp.Body)
 	Expect(err).ToNot(HaveOccurred())
 }
+
+func getNumaflowControllerDefinitions() config.NumaflowControllerDefinitionConfig {
+	// Read definitions config file
+	// TODO: use this file instead "../../tests/config/controller-definitions-config.yaml"
```
Comment (xdevxy) [inline internal/controller/suite_test.go:204]: why not use this file in this PR then?

File: internal/controller/numaflowcontrollerrollout_controller.go [hunk 3]
```diff
@@ -352,38 +368,27 @@ func (r *NumaflowControllerRolloutReconciler) getResourceOperations() (kubeUtil.
...
+	numaflowControllerDeployments := appsv1.Deployment{
+		ObjectMeta: v1.ObjectMeta{Name: NumaflowControllerDeploymentName},
+	}
+	if err := controller.Watch(source.Kind(mgr.GetCache(), &numaflowControllerDeployments),
```
Comment (xdevxy) [inline internal/controller/numaflowcontrollerrollout_controller.go:385]: We may need to watch all related resources defined in [here](https://github.com/numaproj/numaplane/blob/main/config/manager/numaflow-controller-definitions-config.yaml) for Numaflow controller, including configmap.

File: internal/controller/numaflowcontrollerrollout_controller.go (hunk 1 above)
Comment (afugazzotto) [reply internal/controller/numaflowcontrollerrollout_controller.go:381]: I meant that that we are watching the numaflow-controller deployment but we could watch additional resources as needed. I'll change the comment a bit to clarify.

**Review by xdevxy — CHANGES_REQUESTED** [review]

File: internal/controller/suite_test.go (hunk 2 above)
Comment (afugazzotto) [reply internal/controller/suite_test.go:208]: Will leave this for @chandankumar4 PR to avoid conflicts.

**Review by xdevxy — APPROVED** [review]

**Pull request merged**